Add state machine variant with conditional state transition

rpls opened the pull request:
So this is the current WIP for the state machine with conditional transition. I opted for an auxiliary constructor to pass the condition to the state machine, as to make it clear that this will change the behaviour of the fsm. The change works by introducing a "candidate" for `stateNext` named `stateNextCand`. Then `stateNext` will only be assigned this candidate, if the transition condition is met. I chose this design to simplify the treatment of the `onEntry`/`onExitTasks` and to avoid a giant `when()` around these.

I made the synthesis of this completely optional, as to preserve the behaviour of the original state machine. I also introduced a `forceGoto` method for the `StateMachineAccessor`, which sets `stateNext` regardless of the condition. This is, for example, used in the boot state, as I think the boot state should always transition to the entry state regardless of the condition.

Closes #219

rpls commented:
I verified this change on an Ice40 FPGA, and in the toy example the approach saves a bit of logic resources at a very small cost to frequency.

Dolu1990 commented:
hmm maybe,
Maybe, instead of having an alternate  constructor, we could have a function call to set the condition ?

```scala
new StateMachine{
   setTransitionCondition(xxx) ?
}
````

Because ultimately a state machine could have many more things of that kind. 
Maybe that's just taste and flavor XD

But it look good :D

So do you wanna add the function ? or i merge directly ?

Dolu1990 commented:
@rpls

rpls commented:
My initial idea with the alternate constructor was to "express" that this is a different type of statemachine (since it behaves differently). But I think your "speaking" API is better. So I took your suggestion into account. I updated and rebased the branch.
